src/FsAutoComplete/LspServers/FSharpLspClient.fs

```diff
@@ -143,74 +144,164 @@ type ServerProgressReport(lspClient: FSharpLspClient, ?token: ProgressToken) =
-              | false, _ ->
-
-                ()
-            | other -> ()
-
-          message
-      with e ->
-        ()
-
-  member _.DisableEvents(source) = ``base``.DisableEvents(source)
-
-  interface System.IDisposable with
-    member this.Dispose() =
-      lock this
-      <| fun () ->
-        if isNull source then () else this.DisableEvents(source)
-        isDisposed <- true
-        inflightEvents.Values |> Seq.iter (tryDispose)
-        inflightEvents <- null
+  let inflightEvents =
+    ConcurrentDictionary<_, System.Diagnostics.Activity * ServerProgressReport>()
+
+  let isStopped (activity: Activity) =
+#if NET6_0
+    false
+#else
+    activity.IsStopped
+#endif
+    // giving this 1 seconds to report something, otherwise assume it's a dead activity
+    || ((DateTime.UtcNow - activity.StartTimeUtc) > TimeSpan.FromSeconds(1.)
+        && activity.Duration = TimeSpan.Zero)
+
+  let getTagItemSafe key (a: Activity) = a.GetTagItem key |> Option.ofObj
+
+  let getFileName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.fileName
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+  let getProject =
+    getTagItemSafe Tracing.SemanticConventions.FCS.project
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+
+  let getUserOpName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.userOpName
+    >> Option.map string
+    >> Option.defaultValue String.Empty
+
+  let mbp =
+    MailboxProcessor.Start(fun inbox ->
+      async {
+        while not isDisposed do
+          for (a, p) in inflightEvents.Values do
+            // We don't get always get a corresponding ActivityStopped for a given Activity event so this looks at anything that hasn't had a duration incresed at all
+            // We don't seem to get a corresponding ActivityStopped on cancelled typechecks.
+            if isStopped a then
+              do! p.End()
+              inflightEvents.TryRemove(a.Id) |> ignore
+            else
+              // FSC doesn't start their spans with tags so we have to see if it's been added later https://github.com/dotnet/fsharp/issues/14776
+              let fileName = getFileName a
+              let userOpName = getUserOpName a
+              do! p.Report(message = $"{fileName} - {userOpName}")
+
+          match! inbox.TryReceive(250) with
+          | None ->
+            // if we don't get an event in 250 ms just loop again so we can analyze inflightEvents
+            ()
+          | Some(action, activity: Activity, reply: AsyncReplyChannel<unit>) ->
+
+            match action with
+            | "start" ->
+              let fileName = getFileName activity
+              let userOpName = getUserOpName activity
+
+              logger.debug (
+                Log.setMessageI
+                  $"Started : {activity.DisplayName:DisplayName} - {userOpName:UserOpName} - {fileName:fileName}"
```

Comment:
Yes this is using [String interpolation](https://github.com/TheAngryByrd/FsLibLog#string-interpolation). Yes this is still using structured logs.

Reply:
Also these logs aren't getting stuff from tags correctly (as noted above), so it log things like "BackgroundCompiler.Thing - -". Could be more dynamic about building this string so when FCS fixes it, we get these logs correctly but not log blank info.